framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java

```diff
@@ -4631,6 +4613,442 @@ private void makeGroundTargetType(
+   * this method returns {@code type}.
+   *
+   * @param type type to capture
+   * @param typeMirror type from the Java compiler that has captured wildcards, which are used as
+   *     the underlying type of the captured wildcards
+   * @return the captured type
+   */
+  public AnnotatedTypeMirror applyCaptureConversion(
+      AnnotatedTypeMirror type, TypeMirror typeMirror) {
+    if (!shouldCapture(type, typeMirror)) {
+      if (type.containsUninferredTypeArguments()
+          && typeMirror.getKind() == TypeKind.DECLARED
+          && type.getKind() == TypeKind.DECLARED) {
+        // If capture conversion is not applied because the type contains uninferred
+        // wildcards, then mark any uncaptured wildcards as "uninferred".
+        DeclaredType capturedTypeMirror = (DeclaredType) typeMirror;
+        AnnotatedDeclaredType uncapturedType = (AnnotatedDeclaredType) type;
+        for (int i = 0; i < capturedTypeMirror.getTypeArguments().size(); i++) {
+          TypeMirror capturedTypeArgTM = capturedTypeMirror.getTypeArguments().get(i);
+          AnnotatedTypeMirror uncapturedTypeArg = uncapturedType.getTypeArguments().get(i);
+          if (uncapturedTypeArg.getKind() != TypeKind.WILDCARD) {
+            continue;
+          }
+          if (TypesUtils.isCaptured(capturedTypeArgTM)
+              || capturedTypeArgTM.getKind() != TypeKind.WILDCARD) {
+            ((AnnotatedWildcardType) uncapturedTypeArg).setUninferredTypeArgument();
+          }
+        }
+      }
+      return type;
+    }
+
+    DeclaredType capturedTypeMirror = (DeclaredType) typeMirror;
+    AnnotatedDeclaredType uncapturedType = (AnnotatedDeclaredType) type;
+    final AnnotatedDeclaredType capturedType =
+        (AnnotatedDeclaredType) AnnotatedTypeMirror.createType(capturedTypeMirror, this, false);
+
+    captureTypeArgCopier.copy(uncapturedType, capturedType);
+
+    AnnotatedDeclaredType typeDeclaration =
+        (AnnotatedDeclaredType) getAnnotatedType(uncapturedType.getUnderlyingType().asElement());
+
+    // A mapping from type variable to its type argument in the captured type.
+    Map<TypeVariable, AnnotatedTypeMirror> typeVariableMapping = new HashMap<>();
+    // A mapping from the captured type to the annotated captured type.
+    Map<TypeVariable, AnnotatedTypeVariable> capturedTypeMapping = new HashMap<>();
+    List<AnnotatedTypeMirror> newTypeArgs = new ArrayList<>();
+    for (int i = 0; i < typeDeclaration.getTypeArguments().size(); i++) {
+      TypeVariable typeVarTypeMirror =
+          (TypeVariable) typeDeclaration.getTypeArguments().get(i).getUnderlyingType();
+      AnnotatedTypeMirror uncapturedTypeArg = uncapturedType.getTypeArguments().get(i);
+      AnnotatedTypeMirror capturedTypeArg = capturedType.getTypeArguments().get(i);
+      if (uncapturedTypeArg.getKind() == TypeKind.WILDCARD) {
+        if (TypesUtils.isCaptured(capturedTypeArg.getUnderlyingType())) {
```

Comment:
The Javadoc of `TypesUtils.isCaptured` says "Returns whether a TypeVariable represents a captured type." but I think it is actually whether the type variable comes from a wildcard.  The type variable is part of a captured type but is not itself a captured type.  Although javac's `Type.java` poorly uses the name `isCaptured`, I think that TypesUtils should use something more descriptive, such as `isCapturedTypeVariable` or `isTypeVariableFromCaptureConversion` or the like.

Reply:
I replied to this comment https://github.com/typetools/checker-framework/pull/3922/files#r651288918.